core/editor/src/dispatcher/mod.rs

```diff
@@ -13,38 +13,38 @@ impl Dispatcher {
 		log::trace!("{:?}", event);
 		match event {
 			Event::SelectTool(tool_type) => {
-				state.tools.active_tool = tool_type;
+				state.tool_state.active_tool_type = tool_type;
 				Ok(())
 			}
 			Event::SelectPrimaryColor(color) => {
-				state.tools.primary_color = color;
+				state.tool_state.primary_color = color;
 				Ok(())
 			}
 			Event::SelectSecondaryColor(color) => {
-				state.tools.secondary_color = color;
+				state.tool_state.secondary_color = color;
 				Ok(())
 			}
 			Event::SwapColors => {
-				std::mem::swap(&mut state.tools.primary_color, &mut state.tools.secondary_color);
+				std::mem::swap(&mut state.tool_state.primary_color, &mut state.tool_state.secondary_color);
 				Ok(())
 			}
 			Event::ResetColors => {
-				state.tools.primary_color = Color::BLACK;
-				state.tools.secondary_color = Color::WHITE;
+				state.tool_state.primary_color = Color::BLACK;
+				state.tool_state.secondary_color = Color::WHITE;
 				Ok(())
 			}
 			Event::MouseDown(mouse_state) => {
-				state.tools.mouse_state = mouse_state;
-				// the state has changed so we add a trace point
-				state.tools.record_trace_point();
+				state.tool_state.mouse_state = mouse_state;
+				// The state has changed so we add a trace point.
+				state.tool_state.record_trace_point();
 
-				// self.emit_response(Response::UpdateCanvas { document: state.document.render() });
+				state.tool_state.active_tool().handle_input(event);
 				Ok(())
 			}
 			Event::MouseUp(mouse_state) => {
-				state.tools.mouse_state = mouse_state;
-				// the state has changed so we add a trace point
-				state.tools.record_trace_point();
+				state.tool_state.mouse_state = mouse_state;
+				// The state has changed so we add a trace point.
+				state.tool_state.record_trace_point();
 
 				state.document.svg.push(SvgElement::Circle(Circle {
```

Comment:
Could you if you decide that it is feasible to Include the Operations in this PR, move this to the Circle Tool? 
That would be an awesome end to end demonstration and show how the information flow is intended to take place

Reply:
Operations that actually work probably not, I can do that in the next PR though.
I can make an enum though.